glyphs-reader/src/propagate_anchors.rs

```diff
@@ -0,0 +1,883 @@
+    for name in todo {
+        // temporarily remove the glyph so we can mess with it without borrowing all glyphs
+        let glyph = glyphs.get(&name).unwrap();
+        // we work with indices here to also get around borrowck, because
+        // if we borrow the layers in the loop we can't set the anchors in the loop body
+        for layer in &glyph.layers {
+            let anchors = anchors_traversing_components(
+                glyph,
+                layer,
+                glyphs,
+                &all_anchors,
+                &mut num_base_glyphs,
+            );
+            if anchors.len() != layer.anchors.len() {
+                debug_new_anchors(&anchors, glyph, layer);
+            }
+            all_anchors.entry(name.clone()).or_default().push(anchors);
+        }
+    }
+
+    // finally update our glyphs with the new anchors, where appropriate
+    for (name, layers) in all_anchors {
+        let glyph = glyphs.get_mut(&name).unwrap();
+        if glyph.has_components() {
+            assert_eq!(layers.len(), glyph.layers.len());
+            for (i, layer_anchors) in layers.into_iter().enumerate() {
+                glyph.layers[i].anchors = layer_anchors;
+            }
+        }
+    }
+}
+
+fn debug_new_anchors(anchors: &[Anchor], glyph: &Glyph, layer: &Layer) {
+    if !log::log_enabled!(log::Level::Info) {
+        return;
+    }
+    let prev_names: Vec<_> = layer.anchors.iter().map(|a| &a.name).collect();
+    let new_names: Vec<_> = anchors.iter().map(|a| &a.name).collect();
+    log::info!(
+        "propagated anchors for ('{}': {prev_names:?} -> {new_names:?}",
+        glyph.name,
+    );
+}
+
+/// Return the anchors for this glyph, including anchors from components
+fn anchors_traversing_components<'a>(
+    glyph: &'a Glyph,
+    layer: &'a Layer,
+    glyphs: &BTreeMap<SmolStr, Glyph>,
+    // map of glyph -> anchors, per-layer, updated as we do each glyph
+    visited_anchors: &HashMap<SmolStr, Vec<Vec<Anchor>>>,
+    // each (glyph, layer) writes its number of base glyphs into this map during traversal
+    base_glyph_counts: &mut HashMap<(SmolStr, &'a str), usize>,
+) -> Vec<Anchor> {
+    if layer.anchors.is_empty() && layer.components().count() == 0 {
+        return Vec::new();
+    }
+
+    // if this is a mark and it has anchors, just return them.
+    if !layer.anchors.is_empty() && glyph.category == Some(Category::Mark) {
+        let mut anchors = layer.anchors.clone();
+        // if there is an '*origin' anchor, other positions are adjusted by its pos:
+        if let Some(origin) = anchors
+            .iter()
+            .position(Anchor::is_origin)
+            .map(|idx| anchors.remove(idx))
+        {
+            anchors
+                .iter_mut()
+                .for_each(|a| a.pos -= origin.pos.to_vec2())
+        }
+        return anchors;
+    }
+
+    let is_ligature = glyph.sub_category == Some(Subcategory::Ligature);
+    let mut has_underscore = layer
```

Comment:
A name that indicates what has_underscore means would help.

There seems to be quite a lot of anchor name comparison with strings, many of which are repeated ("_", "entry", "exit", etc appear several times). I wonder if we might either make helper functions like `is_exit` or `is_whatever_starts_with_underscore_means` or even something like an AnchorType enum that can be build from the name.

Reply:
- this is the original name, and the original logic, and the original code doesn't really document its intention.
- I tried the helpers, but unfortunately half the places where we're checking these things we're just working with strings, so we can't really add a method without getting fancy.
- the enum thing is nice and we do that once we get to IR, but I think down here we just have to do our best to match the glyphs code without trying to be more systematic than it is.

Reply:
Even if we just made a function it would be an improvement over repeating the same string comparison everywhere.

If we have questions about the intent we can ask Georg, perhaps we should assemble a list of questions?

Reply:
okay, added 'is_exit' and 'is_entry' helpers.


Ultimately here it feels like the implementation is the spec; there are a lot of heuristics and squishy bits here that seem like they're probably the result of trying to get some specific pattern to work. It would be nice if there was something in the glyphs docs that detailed the naming convention for anchors, but for this code in particular I don't have questions, exactly; I'm just taking it at face value and trying to match the behaviour. This is a very literal port.

Reply:
If we don't understand the naming conventions why don't we ask the Glyphs folks? - decent chance they'd be up for explaining.